LUCENE-10357 Ghost fields and postings/points

shahrs87 opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

#11393
Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/lucene/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

shahrs87 commented:
Hi @jpountz,
I have tried to put up a patch from the suggestions you have made in [LUCENE-10357](https://issues.apache.org/jira/browse/LUCENE-10357). Can you please review and provide feedback that I am going in the right direction ? Thank you.

jpountz commented:
Thanks for tackling this! Your are going in the right direction. Could we drop most of the `if (terms != Terms.EMPTY)` checks and handle them like any non-null `Terms` instance?

jpountz reviewed (commented):
One benefit of this change is that fewer null checks would be required. Furthermore, we should not do tests like `if (terms == Terms.EMPTY)`, empty terms should still work with the main code path?

shahrs87 committed:
LUCENE-10357 Ghost fields and postings/points.

shahrs87 commented on lucene/codecs/src/java/org/apache/lucene/codecs/bloom/BloomFilteringPostingsFormat.java:
```diff
@@ -200,8 +200,8 @@ public Terms terms(String field) throws IOException {
         Terms result = delegateFieldsProducer.terms(field);
-        if (result == null) {
-          return null;
+        if (result == null || result == Terms.EMPTY) {
```
Yes, this test case is failing even with this patch: 
`TestMemoryIndexAgainstDirectory#testRandomQueries` 
Reproducible by: `gradlew :lucene:memory:test --tests "org.apache.lucene.index.memory.TestMemoryIndexAgainstDirectory.testRandomQueries" -Ptests.jvms=8 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=B19145C39C34BD03 -Ptests.gui=false -Ptests.file.encoding=UTF-8`

The underlying reader it is using is MemoryIndex#MemoryIndexReader [here](https://github.com/apache/lucene/blob/main/lucene/memory/src/java/org/apache/lucene/index/memory/MemoryIndex.java#L1405)
This is the relevant snippet.
```
        if (info == null || info.numTokens <= 0) {
          return null;
        }
```

Below is the text I copied from LUCENE-10357 description.

> I fear that this could be a source of bugs, as a caller could be tempted to assume that he would get non-null terms on a FieldInfo that has IndexOptions that are not NONE. Should we introduce a contract that FieldsProducer (resp. PointsReader) must return a non-null instance when postings (resp. points) are indexed?

I don't know which all places I need to do null check ? From the above description, looks like only in FieldsProducer related classes. From my limited understanding, this doesn't look like FiledsProducer. @jpountz  please advise.

jpountz replied:
This case is a bit special indeed, but I think we should fix it too to make sure that it only returns a `null` `Terms` instance if the field doesn't exist (fieldInfo == null) or if the field doesn't index terms (indexOptions == NONE).

shahrs87 committed:
more changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
More changes

shahrs87 commented:
@jpountz  Made one more attempt to fix test failures. Can you please take a look again ? Thank you fir being patient with me.

jpountz commented:
Thanks @shahrs87, could you now try to remove all instances of `if (terms == Terms.EMPTY)`? Hopefully existing logic should work with Terms instances regardless of whether they are empty or not.

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -1378,7 +1378,7 @@ private static Status.TermIndexStatus checkFields(
 
       final Terms terms = fields.terms(field);
-      if (terms == null) {
+      if (terms == Terms.EMPTY) {
```
I tried to remove this if (terms == Terms.EMPTY) statement but many tests failed.
Example of failing test: `org.apache.lucene.codecs.lucene90.TestLucene90NormsFormat#testUndeadNorms`
Stack trace:
```
> Task :lucene:core:test FAILED
WARNING: A command line option has enabled the Security Manager
WARNING: The Security Manager is deprecated and will be removed in a future release

org.apache.lucene.codecs.lucene90.TestLucene90NormsFormat > testUndeadNorms FAILED
    org.apache.lucene.index.CheckIndex$CheckIndexException: field "content" should have hasFreqs=true but got false
        at __randomizedtesting.SeedInfo.seed([2A7308C15B316422:27A64CAC8EDF759E]:0)
        at app//org.apache.lucene.index.CheckIndex.checkFields(CheckIndex.java:1437)
        at app//org.apache.lucene.index.CheckIndex.testPostings(CheckIndex.java:2428)
        at app//org.apache.lucene.index.CheckIndex.testSegment(CheckIndex.java:999)
        at app//org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:714)
        at app//org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:552)
        at app//org.apache.lucene.tests.util.TestUtil.checkIndex(TestUtil.java:343)
        at app//org.apache.lucene.tests.store.MockDirectoryWrapper.close(MockDirectoryWrapper.java:909)
        at app//org.apache.lucene.tests.index.BaseNormsFormatTestCase.testUndeadNorms(BaseNormsFormatTestCase.java:698)
        at java.base@17.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base@17.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base@17.0.2/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base@17.0.2/java.lang.reflect.Method.invoke(Method.java:568)
        at app//com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
        at app//com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)
        at app//com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)
        at app//com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)
        at app//org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
        at app//org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at app//org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
        at app//org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at app//org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at app//org.junit.rules.RunRules.evaluate(RunRules.java:20)
```

I think the problem is:
Even though the `content` field is deleted, it is still present in fields returned by `reader#getPostingsReader` [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/CheckIndex.java#L2417) 
Since the returned terms is EMPTY it has `hasFreqs` set to `false` but since this field exists in fields and the indexOptions set to `DOCS_AND_FREQS_AND_POSITIONS` it is expecting hasFreqs to true.

shahrs87 committed:
Add more changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 commented on lucene/core/src/java/org/apache/lucene/index/FrozenBufferedUpdates.java:
```diff
@@ -595,7 +595,7 @@ private void setField(String field) throws IOException {
     DocIdSetIterator nextTerm(String field, BytesRef term) throws IOException {
       setField(field);
-      if (termsEnum != null) {
+      if (termsEnum != null && termsEnum != TermsEnum.EMPTY) {
```
The following test failed `TestPointQueries#testAllPointDocsWereDeletedAndThenMergedAgain`
```
Cannot read field "bytes" because "other" is null
java.lang.NullPointerException: Cannot read field "bytes" because "other" is null
	at __randomizedtesting.SeedInfo.seed([16B9FF96CE2E2AF5:59A84BA7E3282B90]:0)
	at org.apache.lucene.util.BytesRef.compareTo(BytesRef.java:159)
	at org.apache.lucene.index.FrozenBufferedUpdates$TermDocsIterator.nextTerm(FrozenBufferedUpdates.java:604)
	at org.apache.lucene.index.FrozenBufferedUpdates.applyTermDeletes(FrozenBufferedUpdates.java:473)
	at org.apache.lucene.index.FrozenBufferedUpdates.apply(FrozenBufferedUpdates.java:175)
	at org.apache.lucene.index.IndexWriter.forceApply(IndexWriter.java:5965)
	at org.apache.lucene.index.IndexWriter.tryApply(IndexWriter.java:5865)
	at org.apache.lucene.index.IndexWriter.lambda$publishFrozenUpdates$10(IndexWriter.java:2771)
	at org.apache.lucene.index.IndexWriter$EventQueue.processEventsInternal(IndexWriter.java:328)
	at org.apache.lucene.index.IndexWriter$EventQueue.processEvents(IndexWriter.java:317)
	at org.apache.lucene.index.IndexWriter.processEvents(IndexWriter.java:5708)
	at org.apache.lucene.index.IndexWriter.doFlush(IndexWriter.java:4038)
	at org.apache.lucene.index.IndexWriter.flush(IndexWriter.java:3995)
	at org.apache.lucene.index.IndexWriter.forceMerge(IndexWriter.java:2099)
	at org.apache.lucene.index.IndexWriter.forceMerge(IndexWriter.java:2080)
	at org.apache.lucene.search.TestPointQueries.testAllPointDocsWereDeletedAndThenMergedAgain(TestPointQueries.java:1221)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
```

shahrs87 committed:
More changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
More changes

shahrs87 committed:
More changes

shahrs87 committed:
More changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 commented:
> could you now try to remove all instances of if (terms == Terms.EMPTY)?

 @jpountz, I tried to remove all the instances of `if (terms == Terms.EMPTY)?` but couldn't remove the remaining ones in the patch. Otherwise it will cause test failures.

jpountz reviewed (commented):
I left a suggestion as to how to address these failures.

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 commented:
@jpountz  There are no null or terms.EMPTY checks in CheckIndex class anymore.

shahrs87 commented:
@jpountz  Hi Adrian, can you please make one more pass over the PR and provide your feedback ? Thank you.

jpountz commented:
@shahrs87 Can you look into removing all other instances of `terms == Terms.EMPTY` or `terms != Terms.EMPTY` as well? To do this while keeping tests passing, I think you'll need to create empty `Terms` instances that still honor the options of the `FieldInfo` as per my previous suggestion. E.g. you could add a new `Terms#empty(FieldInfo)` helper method that does the right thing for `hasFreqs()`, `hasPositions()`, etc. by looking at the index options of the `FieldInfo`.

shahrs87 committed:
Merge with main branch

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 commented:
Thank you @jpountz  for being so patient with me. I tried your above suggestion and hit the following problem.
Locally I removed the following check from BloomFilteringPostingsFormat.java.
```
        if (result == Terms.EMPTY) {
          return Terms.EMPTY;
        }
```

The following test failed: `org.apache.lucene.index.TestPostingsOffsets.testCrazyOffsetGap`
Can be reproduced by:
 `gradlew :lucene:core:test --tests "org.apache.lucene.index.TestPostingsOffsets.testCrazyOffsetGap" -Ptests.jvms=8 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=66D42010A32F9625 -Ptests.locale=chr -Ptests.timezone=Asia/Jayapura -Ptests.gui=false -Ptests.file.encoding=UTF-8`

The exception stack trace is:
```
field "foo" should have hasFreqs=true but got false
org.apache.lucene.index.CheckIndex$CheckIndexException: field "foo" should have hasFreqs=true but got false
	at __randomizedtesting.SeedInfo.seed([66D42010A32F9625:91A6069AD047326B]:0)
	at app//org.apache.lucene.index.CheckIndex.checkFields(CheckIndex.java:1434)
	at app//org.apache.lucene.index.CheckIndex.testPostings(CheckIndex.java:2425)
	at app//org.apache.lucene.index.CheckIndex.testSegment(CheckIndex.java:999)
	at app//org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:714)
	at app//org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:552)
	at app//org.apache.lucene.tests.util.TestUtil.checkIndex(TestUtil.java:343)
	at app//org.apache.lucene.tests.store.MockDirectoryWrapper.close(MockDirectoryWrapper.java:909)
	at app//org.apache.lucene.index.TestPostingsOffsets.testCrazyOffsetGap(TestPostingsOffsets.java:462)
```

The terms object [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/CheckIndex.java#L1380) is of type [PerFieldPostingsFormat#FieldsReader](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/codecs/perfield/PerFieldPostingsFormat.java#L352)
The fieldsProducer object within  PerFieldPostingsFormat#FieldsReader is of type [BloomFilteringPostingsFormat#BloomFilteredFieldsProducer](https://github.com/apache/lucene/blob/main/lucene/codecs/src/java/org/apache/lucene/codecs/bloom/BloomFilteringPostingsFormat.java#L202)
The delegateFieldsProducer within BloomFilteringPostingsFormat#BloomFilteredFieldsProducer is of type [Lucene90BlockTreeTermsReader](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/Lucene90BlockTreeTermsReader.java#L291)

This is the code snippet which I changed within `Lucene90BlockTreeTermsReader#terms` method
```
  @Override
  public Terms terms(String field) throws IOException {
    assert field != null;
    Terms terms = fieldMap.get(field);
    return terms == null ? Terms.EMPTY : terms;
  }
```

From your suggestion instead of returning Terms.EMPTY, I thought to return Terms.empty(fieldInfo) with overloaded hasFreqs, hasPositions, etc. methods. But the problem is there is no way to get hold of `FieldsInfo` object from `field` string. The fieldMap map within Lucene90BlockTreeTermsReader is empty.  Is it ok to change the method argument for terms method from field String to fieldInfo object within Lucene90BlockTreeTermsReader ?  `public Terms terms(String field) throws IOException` --> `public Terms terms(FieldInfo fieldInfo) throws IOException` I think NO but just wanted to ask.

Please correct me if I am misunderstanding anything. Thank you again.

shahrs87 committed:
Adding Terms#empty(FieldInfo) method

jpountz reviewed (commented):
I wonder if we could fix this by iterating over all fields from field infos in the constructor (which can be accessed through the segment state) and adding an empty instance to the map if there is none already.

shahrs87 committed:
more changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
revert more changes

shahrs87 committed:
Revert "Adding Terms#empty(FieldInfo) method"

This reverts commit 1d3976a9294ff897b5c8e3d2c39ab01506d41104.

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
Changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
more changes

shahrs87 committed:
Merge into main

shahrs87 committed:
New changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
final changes

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10357

shahrs87 committed:
javadoc changes

shahrs87 commented:
I was busy with some other security related work at my day job so couldn't update this PR. Apologies for that.
@jpountz  Can you please review this PR again ?

shahrs87 commented:
@jpountz  Can you please review this patch again? Thank you

mikemccand commented:
Thank you for persisting so hard on this one @shahrs87 -- I'm sorry it looks like we should close it at this point, but your efforts / iterations were needed to see that we are mostly exchanging one `if` for another.

The whole `null` vs `EMPTY` is such a challenge (for Lucene anyways).
